Springboot quickstart web (#2491)
FusionAuth/fusionauth-site: astro/src/content/quickstarts/quickstart-springboot-web.mdx

import RemoteCode from '../../components/RemoteCode.astro';
import SpringBootPreReqs from '../../components/quickstarts/springboot-prereqs.mdx';

In this quickstart, you will build a Spring Boot web application and integrate it with FusionAuth. The application is for [ChangeBank](https://www.youtube.com/watch?v=pkH-kD73QUM), a global leader in converting dollars into coins. It will have areas reserved for logged in users and public-facing sections. It uses Spring Boot web and Thymeleaf, and sets up an OAuth authorization code grant flow, allowing users to log in to the application.

Find the Docker Compose file and source code for the complete application at https://github.com/FusionAuth/fusionauth-quickstart-springboot-web.

## Prerequisites

> <SpringBootPreReqs/>

Reviewer: Should there be a minimum version or at least a 'this quickstart was written with this version of Java'

Reviewer: I see there is a note at the end of the prereqs about Java 17.  That should be moved up to the Java section.

<LoginAfter alt={"Request flow during login after FusionAuth"}/>

> In general, you would introduce FusionAuth to normalize and consolidate user data, making it consistent and up-to-date and offloading your login security and functionality to FusionAuth.

Reviewer: ```suggestion
In general, you would introduce FusionAuth to normalize and consolidate user data, making it consistent and up-to-date. You would also offload your login security and functionality to FusionAuth.
```

## Getting Started

Start with getting FusionAuth up and running and creating a Spring Boot application with Thymeleaf.

### Clone The Code

First, grab the code from the repository and change into that directory.

```
git clone https://github.com/FusionAuth/fusionauth-quickstart-springboot-web.git
cd fusionauth-quickstart-springboot-web
```

### Run FusionAuth Via Docker

You'll find a Docker Compose file `docker-compose.yml` and an environment variables configuration file `.env` in the root directory of the repo.

Assuming you have Docker installed, you can start FusionAuth on your machine with the following.

```

> docker-compose up -d

Reviewer: ```suggestion
docker compose up -d
```

```

This will start three containers, one each for FusionAuth, Postgres, and Elasticsearch.

Here you are using a bootstrapping feature of FusionAuth, called [Kickstart](/docs/v1/tech/installation-guide/kickstart). When FusionAuth starts for the first time, it will look at the `kickstart/kickstart.json` file and configure FusionAuth to your specified state.

> <Aside type="note" >If you ever want to reset the FusionAuth system, delete the volumes created by `docker-compose` by executing `docker-compose down -v`, then rerun `docker-compose up -d`.</Aside>

Reviewer: ```suggestion
<Aside type="note" >If you ever want to reset the FusionAuth system, delete the volumes created by `docker compose` by executing `docker compose down -v`, then rerun `docker compose up -d`.</Aside>
```

FusionAuth will be configured with these settings:

* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `super-secret-secret-that-should-be-regenerated-for-production`.
* Your example username is `richard@example.com` and your password is `password`.
* Your admin username is `admin@example.com` and your password is `password`.
* The base URL of FusionAuth is `http://localhost:9011/`.

> You can log in to the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want to, but with Docker and Kickstart, everything will already be configured correctly.

Reviewer: When I try to login to the admin UI I do not see the lock to log in.
![Screenshot 2023-09-19 at 8 25 02 PM](https://github.com/FusionAuth/fusionauth-site/assets/137117976/c91d2609-dd2b-4e9f-a50e-be8758e42b5b)

Reviewer 2: Fixed in the PR here https://github.com/FusionAuth/fusionauth-quickstart-springboot-web/pull/1

<Aside type="caution">
  The `.env` and `kickstart.json` files contain passwords. In a real application, always add these files to your `.gitignore` file and never commit secrets to version control.
</Aside>

### Create A Basic Spring Boot Application

Next, you'll set up a basic Spring Boot project with a single app. While this guide builds a new Spring Boot project, you can use the same method to integrate your existing project with FusionAuth.

#### Get The Template

> The template is already downloaded in the example repository, but if you wish to download it yourself, go to the Initializr site at https://start.spring.io/ and download your own starter package.

Reviewer: Being new to spring initializr, I'm not sure what settings to use.  It would be useful to add them here.

> You can unzip the package in any directory you would like to. This guide assumes you are in `stub-application` in the example repository, so you should substitute any references to that directory with the directory you unzipped the files in.

Reviewer: Is it a big deal to change from `stub-application` to your-application`?  We have done a few of these now and generally use `your-application`.

Reviewer 2: Made the requested change in the PR here https://github.com/FusionAuth/fusionauth-quickstart-springboot-web/pull/1

> This example was made with Maven, Spring Boot version 2.7.12, and Java 17. If you use different options, the configuration and code may be different.

Reviewer: I think this should be up on the Prereqs.

#### Install Dependencies And Build The Application

This project has three dependencies:

- OAuth2 Client
- Thymeleaf
- Spring Web

From the `stub-application` directory, run the following to execute the Maven build tool to download dependencies and build your application.

```shell

> ./mvnw package

Reviewer: I am not familiar with Java and when I try to run this, I get the following error:

`./mvnw package
./mvnw: line 280: /Users/mark.robustelli/Projects/quickstart-springboot-web/fusionauth-quickstart-springboot-web/stub-application/.mvn/wrapper/maven-wrapper.properties: No such file or directory
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Failed to open the file 
Warning: /Users/mark.robustelli/Projects/quickstart-springboot-web/fusionauth-q
Warning: uickstart-springboot-web/stub-application/.mvn/wrapper/maven-wrapper.j
Warning: ar: No such file or directory
 11 58727   11  6886    0     0  15063      0  0:00:03 --:--:--  0:00:03 15134
curl: (23) Failure writing output to destination
Error: Could not find or load main class org.apache.maven.wrapper.MavenWrapperMain
Caused by: java.lang.ClassNotFoundException: org.apache.maven.wrapper.MavenWrapperMain
`
Am I missing something or is there something I need to do to work around this?

* Integrate with external systems using [Webhooks](/docs/v1/tech/events-webhooks/), [SCIM](/docs/v1/tech/core-concepts/scim), and [Lambdas](/docs/v1/tech/lambdas/).

## Troubleshooting

* I get "This site can’t be reached localhost refused to connect" when I click the login button.

Ensure FusionAuth is running in the Docker container. You should be able to log in as the admin user, `admin@example.com` with a password of `password` at [http://localhost:9011/admin](http://localhost:9011/admin).

* I get an error page when I click on the login button with the message `"error_reason" : "invalid_client_id"`.

Ensure the value for `spring.security.oauth2.client.registration.fusionauth-client.client-id` in the file `src/main/resources/application.properties` matches client the Id configured in FusionAuth for the "Example App" Application at [http://localhost:9011/admin/application/](http://localhost:9011/admin/application/).

* It still doesn't work.

You can always pull down a complete running application and compare what's different.

```
git clone https://github.com/FusionAuth/fusionauth-quickstart-springboot-web.git
cd fusionauth-quickstart-springboot-web

> docker-compose up -d

Reviewer: ```suggestion
docker compose up -d
```

* The base URL of FusionAuth is `http://localhost:9011/`.

You can log in to the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want to, but with Docker and Kickstart, everything will already be configured correctly.

<Aside type="caution">
  The `.env` and `kickstart.json` files contain passwords. In a real application, always add these files to your `.gitignore` file and never commit secrets to version control.
</Aside>

### Create A Basic Spring Boot Application

Next, you'll set up a basic Spring Boot project with a single app. While this guide builds a new Spring Boot project, you can use the same method to integrate your existing project with FusionAuth.

#### Get The Template

Go to the Spring Initializr site at https://start.spring.io/ and download your own starter package. Choose the following options.

* Project - `Mavern`
* Language - `Java`
* Spring Boot Version - `2.7.16`
* Project Metadata
    * Group - `io.fusionauth.quickstart`
    * Artifact - `springweb`
    * Package Name - `io.fusionauth.quickstart.springweb`
    * Packaging - `Jar`
    * Java Version - `17`

> This project has the following three dependencies which you will need to add by clicking the "Add Dependencies" button when creating the package on the Initializr site.

Reviewer 2: @mark-robustelli can you confirm if you added these dependencies 

![CleanShot 2023-10-06 at 00 27 52@2x](https://github.com/FusionAuth/fusionauth-site/assets/31291528/588e890d-6fb6-46e4-829a-39dae579d259)

The error log you posted shows that the package that you downloaded on the Initializr site does not have the three dependencies listed here

First, `SecurityConfiguration.java`, which contains the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/config/SecurityConfiguration.java"
            lang="java" />

This class uses the `@Configuration` annotation to provide additional configuration, in this case, a `SecurityFilterChain` that tells Spring to:

- Require authentication for all requests *except* to `/` (`home.html`), `/images`, and `/css` (implicitly referring to the directories under `src/main/resources/static`).
- Redirect the user on logout to FusionAuth to complete the logout.

Second, `WebConfig.java`, which contains the code below.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/config/WebConfig.java"
            lang="java" />

This configuration class enables Web MVC and sets up resource handlers for the static assets.

## Customization

> The template files you added to the `templates` directory reference a stylesheet and images. Create a `css` and `images` directory under `src/main/resources/static/` to place the stylesheet and images.

Reviewer 2: Please see my comment https://github.com/FusionAuth/fusionauth-site/pull/2491#discussion_r1348076056 if the correct dependencies are added when creating the package on the Initializr site the `static` and `templates` directories will already be available
![CleanShot 2023-10-06 at 01 45 21@2x](https://github.com/FusionAuth/fusionauth-site/assets/31291528/c44104f9-0415-4fca-b511-88169ff0afc8)

FusionAuth will be configured with these settings:

* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `super-secret-secret-that-should-be-regenerated-for-production`.
* Your example username is `richard@example.com` and your password is `password`.
* Your admin username is `admin@example.com` and your password is `password`.
* The base URL of FusionAuth is `http://localhost:9011/`.

You can log in to the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want to, but with Docker and Kickstart, everything will already be configured correctly.

<Aside type="caution">
  The `.env` and `kickstart.json` files contain passwords. In a real application, always add these files to your `.gitignore` file and never commit secrets to version control.
</Aside>

### Create A Basic Spring Boot Application

Next, you'll set up a basic Spring Boot project with a single app. While this guide builds a new Spring Boot project, you can use the same method to integrate your existing project with FusionAuth.

> 

Reviewer: ```suggestion

<Aside type="note">
If you simply want to run the application , there is a completed version in the 'complete-application' directory. You can use the following commands to get it up and running if you do not want to create your own.

```shell
cd complete-application
./mvnw package
./mvnw spring-boot:run
```

Then view the website at the following location `http://localhost:8080`.
</Aside>
```

> 

Reviewer: Adding this because we have it in other quick starts.  The comments in git through this off so please make sure the "</Aside>" tag is included.

```suggestion

<Aside type="note">
If you simply want to run the application , there is a completed version in the 'complete-application' directory. You can use the following commands to get it up and running if you do not want to create your own.

```shell
cd complete-application
./mvnw package
./mvnw spring-boot:run
```

Then view the website at the following location `http://localhost:8080`.
</Aside>
```

#### Get The Template

Go to the Spring Initializr site at https://start.spring.io/ and download your own starter package. Choose the following options.

* Project - `Mavern`
* Language - `Java`
* Spring Boot Version - `2.7.16`
* Project Metadata
    * Group - `io.fusionauth.quickstart`
    * Artifact - `springweb`
    * Package Name - `io.fusionauth.quickstart.springweb`
    * Packaging - `Jar`
    * Java Version - `17`

This project has the following three dependencies which you will need to add by clicking the <Uielement label="Add Dependencies"/> button and searching for each dependency, when creating the package on the Initializr site.

- OAuth2 Client
- Thymeleaf
- Spring Web

> If you choose different options, the configuration and code may be different. After entering the configuration settings above and adding the dependencies you can click the <Uielement label="Generate"/> button to download your starter package.

Reviewer: The note to run all command from the your-application should be moved to here.

The properties with the `spring.security.oauth2.provider` prefix tell Spring where the important endpoints for FusionAuth are.

### Add Controllers

Find the base directory where the main application Java class is defined. In our example, the main class is `FusionAuthSpringApplication.java` in `src/main/java/io/quickstart/springweb`.

<Aside type="note">If you downloaded your own starter, you will need to update the `package` for the code snippets below to match your own.</Aside>

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/FusionAuthSpringApplication.java"
            lang="java"/>

Create the following `HomeController.java` Java class in the `src/main/java/io/quickstart/springweb` directory.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/HomeController.java"
            lang="java"/>

> This controller serves the page for unauthenticated users on the `/` route and shows the application name, which is injected from the `spring.application.name` property in the `application.properties` file. The `home` return string tells Spring to render the `home.html` template.

Reviewer: where is the home.html file located?  I see that this created later, so you will want to say something like "...`home.htm` template which will be created later."

> Create the following `AccountController.java` Java class in the `src/main/java/io/quickstart/springweb` directory.

Reviewer: ```suggestion
Create another file in the `src/main/java/io/quickstart/springweb` directory. This time, name it `AccountController.java` and paste the following code into it.
```

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/AccountController.java"
            lang="java"/>

> This controller serves the page for authenticated users on the `/account` route and adds the claims from the `OidcUser` returned from FusionAuth to the view model on the `profile` attribute. The `account` return string tells Spring to render the `account.html` template.

Reviewer: Where is the account.html file?

> Create the following `MakeChangeController.java` Java class in the `src/main/java/io/quickstart/springweb` directory.

Reviewer: ```suggestion
Next, create a file named `MakeChangeController.java` in the `src/main/java/io/quickstart/springweb` directory.  Then, copy the following code into the newly created file.
```

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/MakeChangeController.java"
            lang="java"/>

This controller serves both GET and POST requests on the `/make-change` route. Both methods add the claims to the `profile` attribute like `AccountController.java` does. `GetMapping` adds a new model object of `Change` to the view model. `PostMapping` takes the `Change` object as a parameter and uses the `total` to calculate how many nickels and pennies to return. If `total` is not a valid dollar (`BigDecimal`) value, it adds an error to the `Change` model object. The return string `make-change` tells Spring to render the `make-change.html` template.

### Add The Model Object

> Now you'll create the model object referenced in the controllers. In the project base directory `src/main/java/io/quickstart/springweb`, create a new directory called `model` and add a new `Change.java` class to it.

Reviewer: ```suggestion
Now you'll create the model object referenced in the controllers. In the project base directory `src/main/java/io/quickstart/springweb`, create a new directory called `model`. In the `model` directory, and create a new file named `Change.java`. Now, populate it with the follow code for the `Change` class.
```

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/model/Change.java"
            lang="java" />

This model has four fields.
- `error`: The error string if the input was invalid.
- `total`: The total amount to divide into pennies and nickels.
- `nickels`: The number of nickels returned from the `make change` operation.
- `pennies`: The number of pennies returned from the `make change` operation.

The model also has a `getMessage` method, which formats a message string to render on the template.

### Add View Templates

Now you can add three views for Thymeleaf to render on the routes you defined:

- `Home`: The welcome page an unauthenticated user will see when they visit your application.
- `Account`: A protected view the user will see after they log in.
- `Make Change`: A protected view to demonstrate making an authenticated form post to the backend.

> Create the template files in `src/main/resources/templates`.

Reviewer: Let them know they will need to create the templates folder.

Reviewer 2: I have made adjustments to the instructions. When you download the starter package if you have added the correct dependencies `Thymeleaf` and `Spring Web`, on the Spring Initializr site, the template folder will already be there in your package.

This template links to `/account` for login. Spring will automatically know that the user is not logged in and redirect them to FusionAuth based on what was defined in `application.properties`.

Create the `account.html` template file.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/resources/templates/account.html"
            lang="html" />

This page has an `a href` link to `/logout` that Spring will use to redirect the user to FusionAuth for logout. It also references the `${profile.email}` claim that was set in the account controller.

Create the `make-change.html` template file.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/resources/templates/make-change.html"
            lang="html" />

In addition to the references made above, this template adds a form that binds the `${change}` object as the model, submits the `*{total}` as input by the user, and displays the message or error depending on what is defined in the `Change` object.

### Configure Spring Web And Spring Security

> From the base directory `src/main/java/io/quickstart/springweb`, add a new directory called `config` and add the following two Java classes to it.

Reviewer: ```suggestion
From the base directory `src/main/java/io/quickstart/springweb`, add a new directory called `config` and add the following two files to it.
```

> First, `SecurityConfiguration.java`.

Reviewer: First, `SecurityConfiguration.java`, which contains the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-springboot-web/main/complete-application/src/main/java/io/fusionauth/quickstart/springweb/config/SecurityConfiguration.java"
            lang="java" />

This class uses the `@Configuration` annotation to provide additional configuration, in this case, a `SecurityFilterChain` that tells Spring to:

- Require authentication for all requests *except* to `/` (`home.html`), `/images`, and `/css` (implicitly referring to the directories under `src/main/resources/static`).
- Redirect the user on logout to FusionAuth to complete the logout.

> Second, `WebConfig.java`.

Reviewer: ```suggestion
Second, `WebConfig.java`, which contains the code below.
```
